### ISSUE TITLE
PyInstaller fails with editable `pyproject.toml` installs

### ISSUE THREAD

**Issue by mhils** [pyinstaller/pyinstaller#7524]
After transitioning from `setup.py` to `pyproject.toml`, our PyInstaller builds started to fail with the following traceback:
```
$ ./mitmproxy
Traceback (most recent call last):
  File "mitmproxy", line 1, in <module>
ModuleNotFoundError: No module named 'mitmproxy'
[2041] Failed to execute script 'mitmproxy' due to unhandled exception!
```

I discovered I can fix this by using `pip install .` instead of `pip install -e .` for my package. It seems like PyInstaller has troubles detecting/resolving editable `pyproject.toml` installs. 

I've created a minimal reproducer here: https://github.com/mhils/pyinstaller-pyproject-editable

### Context information (for bug reports)

* Output of `pyinstaller --version`: 5.9.0 and 1c6aa7413d4012f950a987860724f4b4bdcefbd0
* Version of Python: 3.11
* Platform: Linux and Windows
* How you installed Python: deadsnakes / python.org/downloads.

refs https://github.com/mitmproxy/mitmproxy/discussions/6025

**Comment by bwoodsend** [comment]
Ughh. For pure pyproject.toml editable installs, instead of just creating a `.pth` file in site-packages which adds the project's parent directory to `sys.path`, it generates an elaborate *meta path* finder. e.g. Mine are at:
```
~/.pyenv/versions/3.10.9/lib/python3.10/site-packages/__editable__.my_module-1.0.pth
~/.pyenv/versions/3.10.9/lib/python3.10/site-packages/__editable___my_module_1_0_finder.py
```
I dread to think what we'd have to do in order to support this. I was rather hoping that someone at PyPA would realise just how stupid this is and go back to the original behaviour.

**Comment by rokm** [comment]
Suppose we'll see how the `modulegraph` replacement deals with meta path finders, and try to sort this out once we make the transition. As far as my understanding of it goes, the current `modulegraph` approach is oblivious to meta path finders...

**Comment by Legorooj** [comment]
The experimental modulegraph I wrote should, in theory, support this, as it offloads all work to importlib and friends which should support all of this out of the box.

**Comment by maddyaby** [comment]
I'm hitting this same issue. I take it @Legorooj  that your importlib-based modulegraph didn't go in yet?

I added an attempt to use importlib at the end of _find_module_path in modulegraph.py (https://github.com/maddyaby/pyinstaller/commit/bcf642e528511fafc17b0ffb14bfbaf0333f90f5).

This seems to fix my case but might not be the best or cleanest way to do this... any thoughts?

**Comment by Legorooj** [comment]
No, the mg upgrades haven't happened yet. I've got a mostly functional dev version somewhere but I didn't have time to finish it.

Hopefully I can finish it and submit it in the next month or two.

**Comment by n3bul4** [comment]
Hi everyone,

I encountered the same issue where switching from `setup.py` to `pyproject.toml` broke the binary build process for editable modules. While I couldn’t find a direct solution mentioned in the discussion, I stumbled upon a working workaround that ensures editable modules also build correctly as binaries.

The solution involves adding the following configuration to every requirement in editable mode:

`--config-settings editable_mode=compat`

For example, this is how you would apply it:

`--config-settings editable_mode=compat -e modulename`

By adding this to the editable requirements, the build process works correctly for both editable modules and binary distributions.

Hopefully, this can help others who are facing the same issue.

**Comment by mj2068** [comment]
@n3bul4 awesome. mine did work. but i did not find much relevant info about the mechanism why it worked, for example, regarding what `editable_mode=compat` actually does or other `--config-settings` key-value settings.
nonetheless, thank you.
